community/rfcs/techniques/design/20200306-weight-clustering-design.md

```diff
@@ -0,0 +1,144 @@
+    clustered_model = cluster.cluster_weights(model, **clustering_params)
+    ```
+
+    NOTE: Clustering happens as part of the model building process, therefore in order to cluster sequential models, we need to call `model.build()` explicitly after applying `cluster_weights()` &mdash; see [the example](#example-0-cluster-an-entire-model) below.
+
+2. **Fine-tuning** &mdash; train the clustered model to fine-tune the initial values of cluster centroids:
+
+   ```python
+   clustered_model.fit(x_training_data, y_training_data, epochs=1)
+   ```
+
+3. **Stripping** &mdash; call `strip_clustering()` to strip the clustering-specific information from the model and re-obtain the original model structure:
+
+   ```python
+   stripped_model = cluster.strip_clustering(clustered_model)
+   ```
+
+The code snippets below illustrate common usage patterns of the clustering API. For the sake of brevity, the clustering examples presented here only deal with the 1st step of the above workflow.
+
+A more comprehensive end-to-end example on how to use the API to cluster and save a model is provided in the examples directory, under *clustering*.
+
+### Cluster an Entire Model
+
+When passing an entire model to `cluster_weights()`, the tool will determine which layers are clusterable and will cluster all of them &mdash; other layers will be ignored and left as they are.
+
+```python
+clustering_params = {
+    'number_of_clusters': 8,
+    'cluster_centroids_init': 'linear'
+}
+
+# Clustering a sequential model
+model = keras.Sequential([
+    layers.Dense(10, input_shape=(10,)),
+    layers.Dense(10),
+])
+
+clustered_model = cluster.cluster_weights(model, **clustering_params)
+clustered_model.build(input_shape=(1, 10))
+```
+
+```python
+# Clustering a functional model
+i1 = keras.Input(shape=(10,))
+i2 = keras.Input(shape=(10,))
+x1 = layers.Dense(10)(i1)
+x2 = layers.Dense(10)(i2)
+outputs = layers.Add()([x1, x2])
+
+model = keras.Model(inputs=[i1, i2], outputs=outputs)
+
+clustered_model = cluster.cluster_weights(model, **clustering_params)
+```
+
+### Cluster Some Layers
+
+Individual layers within a model can be clustered selectively by invoking `cluster_weights()` on a layer object, as follows:
+
+```python
+clustered_model = keras.Sequential()
+clustered_model.add(cluster.cluster_weights(keras.layers.Dense(10), **clustering_params))
+clustered_model.add(keras.layers.Dense(10))
+```
+In the above example model the weights of the 1st `Dense` layer will be clustered, whereas the weights of the 2nd layer will not be clustered.
+
+### Export Clustered Model
+
+```python
+clustered_model = cluster_weights(model, **clustering_params)
+keras.models.save_model(clustered_model, keras_file)
+```
+
+### Import Clustered Model
+
+A clustered Keras model needs to be loaded within a `cluster_scope` in order to be successfully deserialized:
+
+```python
+with cluster_scope():
+  loaded_model = keras.models.load_model(keras_file)
+```
+
+## Future Improvements
```

Comment:
Please refactor this into the "Compatibility" section from the pruning RFC. 

Both at launch and roadmap. This content would be in the roadmap.